mlir/lib/Dialect/Tensor/IR/TensorOps.cpp

```diff
@@ -4337,11 +4337,16 @@ LogicalResult PackOp::canonicalize(PackOp packOp, PatternRewriter &rewriter) {
       dest =
           rewriter.create<tensor::CastOp>(loc, newDestType, packOp.getDest());
     }
-    Value newOp = rewriter.create<tensor::PackOp>(
-        loc, source, dest, packOp.getInnerDimsPos(), packOp.getMixedTiles(),
-        packOp.getPaddingValue(), packOp.getOuterDimsPerm());
+    auto clonedPackOp = cast<PackOp>(rewriter.clone(*packOp));
```

Comment:
Sorry didnt get to review before it landed but you could use
```
auto discardableAttributes = getPrunedAttributeList(packOp, PackOp::getAttributesList());
clonedPackOp.setAttrs(discardableAttributes);
```

Reply:
Thanks for the suggestion. Looks clean!

Reply:
The `Operation::getDiscardableAttrDictionary()` method seems like a more direct implementation than `getPrunedAttributeList()` (why isn't this one living in mlir/IR by the way??)

That said, I didn't notice in the review that you're cloning the op here: why is that? Why aren't you just modifying it in place? Since you were doing `rewriter.startOpModification` I was assuming that this is what is happening.
When we can avoid recreating an operation and destroying the original, it's just more efficient.

Reply:
Could you suggest how to modify it inplace without cloning in this case?

Reply:
Shortest path should be:
```
  rewriter.modifyOpInPlace(packOp, [&] {
     packOp.getSourceMutable().assign(source);
     packOp.getDestMutable().assign(dest);
     packOp.getResult();.setType(dest.getType());
 });
```

Reply:
The problem was adding `tensor.cast` after the modification. https://github.com/llvm/llvm-project/blob/a649e8ff891e54a279768a8e877b76c8f000a5a7/mlir/lib/Dialect/Tensor/IR/TensorOps.cpp#L4348 I can't replace it with `tensor.pack`.

Reply:
I think it is not trivial to do in-place modification. There are two situations:

1. The dest tensor type is the same. In this case, we do not need a tensor.cast consumer.
2. The dest tensor type is changed. In this case, we need to create the tensor.cast op which makes the types consistent.

In the (1) situation, we can do in-place modification -- which is very simple.

In the (2) situation, it is not trivial because you need to replace the original op with the new tensor.cast op. If we do in-place modification, I don't see a trivial way to replace the op. Perhaps we can replace the uses of the tensor.pack ops with the new tensor.cast op, when it is the case.

IMO, cloning an op is cheap in this case. Instead of adding complex to logics, I'm +1 on cloning the op approach. 

Note that this is also what we're doing for LinalgOps and it's been there for a couple years. I'm not saying that this is the correct way, but it's more like providing data points.

https://github.com/llvm/llvm-project/blob/9f3c55954eaa71910caa0abbb404db02d3a104c1/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp#L2561-L2575

Reply:
> There are two situations:
> - The dest tensor type is the same. In this case, we do not need a tensor.cast consumer.
> - The dest tensor type is changed. In this case, we need to create the tensor.cast op which makes the types consistent.

The current code already does not try differentiate between these I believe: it always creates the cast, which is folded later if the types were matching.

> I think it is not trivial to do in-place modification.

I don't understand what's not trivial here.
I sent a PR implementing it: https://github.com/llvm/llvm-project/pull/111593